app/controllers/application_controller.rb

```diff
@@ -5,6 +5,142 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions unless test?
+    set :session_secret, "my_super_secret_session"
   end
 
-end
\ No newline at end of file
+  helpers do
+    def current_user(session)
+      @user = User.find(session[:id])
+      @user
+    end
+
+    def is_logged_in?(session)
+      !!session[:id] ? true : false
+    end
+  end
+
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if is_logged_in?(session)
+      redirect 'tweets'
+    else
+      erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do
+    @user = User.new(params)
+    if @user.save
+      session[:id] = @user.id
+      redirect 'tweets'
+    else
+      redirect 'signup'
+    end
+  end
+
+  get '/login' do
+    if is_logged_in?(session)
+      redirect 'tweets'
+    else
+      erb :'users/login'
+    end
+  end
+
+  post '/login' do
+    @user = User.find_by(username: params[:username])
+    if !!@user
+      session[:id] = @user[:id]
+      redirect 'tweets'
+    else
+      redirect 'login'
+    end
+  end
+
+  get '/tweets' do
+    if is_logged_in?(session)
+      @user = current_user(session)
+      @tweets = Tweet.all
+      erb :'tweets/tweets'
+    else
+      redirect 'login'
+    end
+  end
+
+  post '/tweets' do
+    if is_logged_in?(session)
+      tweet = Tweet.new(content: params[:content], user_id: session[:id])
+      if tweet.save
+        redirect 'tweets'
+      else
+        redirect 'tweets/new'
+      end
+    else
+      redirect 'login'
+    end
+  end
+
+  get '/tweets/new' do
+    if is_logged_in?(session)
+      erb :'tweets/create_tweet'
+    else
+      redirect 'login'
+    end
+  end
+
+  get '/tweets/:id' do
+    if is_logged_in?(session)
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/show_tweet'
+    else
+      redirect 'login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if is_logged_in?(session) # && session[:id] == @tweet.user_id
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/edit_tweet'
+    else
+      redirect 'login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    if params[:content] == ""
+      redirect "/tweets/#{params[:id]}/edit"
+    else
+      @tweet = Tweet.find(params[:id])
+      @tweet.update(content: params[:content])
+    end
+    redirect "/tweets/#{@tweet.id}"
+  end
+
+  get '/logout' do
+    if is_logged_in?(session)
+      session.clear
+      redirect 'login'
+    else
+      redirect '/'
+    end
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+  delete '/tweets/:id/delete' do
+    @tweet = Tweet.find(params[:id])
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.